core/src/main/java/hudson/model/Run.java

```diff
@@ -1930,31 +1921,55 @@ public synchronized void save() throws IOException {
      * @throws IOException If there is a problem reading the log file.
      */
     public @Nonnull List<String> getLog(int maxLines) throws IOException {
-        int lineCount = 0;
-        List<String> logLines = new LinkedList<String>();
-        if (maxLines == 0) {
-            return logLines;
-        }
-        try (BufferedReader reader = new BufferedReader(new InputStreamReader(new FileInputStream(getLogFile()), getCharset()))) {
-            for (String line = reader.readLine(); line != null; line = reader.readLine()) {
-                logLines.add(line);
-                ++lineCount;
-                // If we have too many lines, remove the oldest line.  This way we
-                // never have to hold the full contents of a huge log file in memory.
-                // Adding to and removing from the ends of a linked list are cheap
-                // operations.
-                if (lineCount > maxLines)
-                    logLines.remove(0);
+        final List<String> lastLines = new ArrayList<>(maxLines);
```

Comment:
🐛 if somebody is dare enough to pass `MAX_INT` in his code, it will consume too much memory. I would limit it by something reasonably small (e.g. 1024) if you really need an ArrayList